### PR TITLE
Add alternative tags implementation

### PR DESCRIPTION
**Pull request by hiqsol**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

This PR needs https://github.com/yiisoft/factory/pull/47 for tests to pass.

### CODE REVIEW

**Comment by BoShurik** [comment]
In this implementation you must manually collect all tagged services. See https://github.com/yiisoft/di/issues/125#issuecomment-645310275

**Comment by BoShurik** [comment]
BTW it works out of the box:
```php
$container = new Container([
    'engines' => [
        '__class' => ArrayIterator::class,
        '__construct()' => [
            [Reference::to(Foo::class), Reference::to(Bar::class)],
            0
        ],
    ],
]);

var_dump($container->get('engines'));
```

**Comment by hiqsol** [comment]
@BoShurik the PR was refactored, added support for same syntax and tests as original implementation

File: src/CompositeContainer.php [hunk 1]
```diff
@@ -29,6 +32,18 @@ public function get($id)
...
+    private function collectTags(string $id): array
+    {
+        $res = [];
+        foreach ($this->containers as $container) {
+            if ($container->has($id)) {
+                $res = array_merge($res, $container->get($id));
+            }
+        }
+
+        return $res;
```
Comment (samdark) [inline src/CompositeContainer.php:44]:
```suggestion
        $results = [];
        foreach ($this->containers as $container) {
            if ($container->has($id)) {
                $results = array_merge($results, $container->get($id));
            }
        }

        return $results;
```

File: tests/Unit/TagsTest.php [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+use Yiisoft\Factory\Exceptions\InvalidConfigException;
+
+/**
+ * ContainerTest contains tests for \Yiisoft\Di\Container
+ */
+class TagsTest extends TestCase
+{
+    public function testTagNameAlreadyInUse(): void
+    {
+        $this->expectException(InvalidConfigException::class);
```
Comment (samdark) [inline tests/Unit/TagsTest.php:26]: Why?

Comment (hiqsol) [reply tests/Unit/TagsTest.php:26]: Cause `engine` is already taken for `EngineMarkOne`
Missed hash sign could be a typo.

It is another question should we allow names without hash for tags.
This implementation mostly doesn't care about # in tag names. 
The hash sign is only needed for `CompositeContainer`.

Comment (samdark) [reply tests/Unit/TagsTest.php:26]: We should not allow non-hashed names, I think.

**Comment by samdark** [comment]
Closing in favor of https://github.com/yiisoft/di/pull/144

**Pull request closed without merging**